homeassistant/components/amberelectric/services.py

```diff
@@ -0,0 +1,107 @@
+"""Amber Electric Service class."""
+
+from typing import Any, cast
+
+from amberelectric.models.channel import ChannelType
+
+from homeassistant.config_entries import ConfigEntryState
+from homeassistant.core import (
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+)
+from homeassistant.exceptions import ServiceValidationError
+
+from .const import ATTR_CHANNEL_TYPE, ATTR_SITE_ID, DOMAIN, GET_FORECASTS_SERVICE
+from .coordinator import AmberConfigEntry
+from .formatters import normalize_descriptor
+
+
+def format_cents_to_dollars(cents: float) -> float:
+    """Return a formatted conversion from cents to dollars."""
+    return round(cents / 100, 2)
+
+
+def async_get_entry(hass: HomeAssistant, site_id: str) -> AmberConfigEntry:
+    """Get the Amber config entry."""
+    if not (entry := hass.config_entries.async_get_entry(site_id)):
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="integration_not_found",
+            translation_placeholders={"target": DOMAIN},
+        )
+    if entry.state is not ConfigEntryState.LOADED:
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="not_loaded",
+            translation_placeholders={"target": entry.title},
+        )
+    return cast(AmberConfigEntry, entry)
+
+
+def get_forecasts(channel_type: str, data: dict) -> list[Any]:
+    """Return an array of forecasts."""
+    results = []
+    if channel_type not in data["forecasts"]:
+        raise ServiceValidationError(
+            translation_domain=DOMAIN,
+            translation_key="channel_not_found",
+            translation_placeholders={"channel_type": channel_type},
+        )
```

Comment:
when would this happen?

Reply:
Not all support all of the main three channel types, ie. it everyone has a `controlled_load`.

Reply:
Also, what is stopping someone using an invalid string in an automation?

Reply:
I believe the validation stops people from doing that

Reply:
Only if they use UI mode. You can do what ever you want in YAML mode. I think capturing and returning a meaningful error he is a better experience, even if it's a rare occurrence.

Reply:
Is there a way to do dynamic validation depending on site configuration?

Reply:
Nope, you didn't set a `schema` on the service registration.

Dynamic validation mean raising ServiceValidationError

Reply:
Let me see if I can work that out. But how do I make that be reflected in the action UI?

Reply:
Oh right, dynamic validation is not dynamic dynamic, but that is the way to tell the users they cant do something

Reply:
Ok, then we still need this check.